contracts/test/signatureBased.test.ts

```diff
@@ -0,0 +1,289 @@
+
+import dotenv from "dotenv";
+import { _TypedDataEncoder } from "ethers/lib/utils";
+dotenv.config();
+
+const PRIVATE_KEY =
+  process.env.WALLET_PRIVATE_KEY ||
+  "0x7726827caac94a7f9e1b160f7ea819f172f7b6f9d2a97f992c38edeab82d4110";
+const abiCoder = new ethers.utils.AbiCoder();
+
+describe("SignatureBasedPaymaster", function () {
+  let provider: Provider;
+  let wallet: Wallet;
+  let deployer: Deployer;
+  let userWallet: Wallet;
+  let signerWallet: Wallet;
+  let paymaster: Contract;
+  let greeter: Contract;
+
+  before(async function () {
+    const deployUrl = hardhatConfig.networks.zkSyncInMemory.url;
+    [provider, wallet, deployer] = setupDeployer(deployUrl, PRIVATE_KEY);
+    const emptyWallet = Wallet.createRandom();
+    console.log(`User wallet's address: ${emptyWallet.address}`);
+    userWallet = new Wallet(emptyWallet.privateKey, provider);
+    signerWallet = Wallet.createRandom();
+    console.log(`Signer wallet's address: ${signerWallet.address}`);
+    signerWallet = new Wallet(signerWallet.privateKey, provider);
+    paymaster = await deployContract(deployer, "SignatureBasedPaymaster", [
+      signerWallet.address,
+    ]);
+    greeter = await deployContract(deployer, "Greeter", ["Hi"]);
+    await fundAccount(wallet, paymaster.address, "3");
+    console.log(`Paymaster current signer: ${signerWallet.address}`);
+  });
+
+  async function createSignatureData(
+    signer: Wallet,
+    user: Wallet,
+    delay: number,
+  ) {
+    const nonce = await paymaster.nonces(user.address);
+    const typeHash = await paymaster.SIGNATURE_TYPEHASH();
+    const eip712Domain = await paymaster.eip712Domain();
+    const currentTimestamp = (await provider.getBlock("latest")).timestamp;
+    const lastTimestamp = currentTimestamp + delay; // 300 seconds
+
+    const domain = {
+      name: eip712Domain[1],
+      version: eip712Domain[2],
+      chainId: eip712Domain[3],
+      verifyingContract: eip712Domain[4],
+    };
+    // @dev : Upon changing types here, you need to ensure that
+    // `SIGNATURE_TYPEHASH` constant in signatureBasedPaymaster contract matches the changes
+    // Otherwise EIP712Domain based _signedTypedData will return wrong signatures.
+    // And test will fail.
+    const types = {
+      SignatureBasedPaymaster: [
+        { name: "userAddress", type: "address" },
+        { name: "lastTimestamp", type: "uint256" },
+        { name: "nonces", type: "uint256" },
+      ],
+    };
+    const values = {
+      userAddress: user.address,
+      lastTimestamp: lastTimestamp,
+      nonces: nonce,
+    };
+
+    const signature = await signer._signTypedData(domain, types, values);
+
+    //console.log("Signature generated successfully !");
+    //console.log("Signer : " + signer.address);
+    //console.log("Paymaster : "+ paymaster.address);
+    //console.log(`Signature valid till: ${lastTimestamp} for user: ${user.address} and nonce: ${nonce}`);
```

Comment:
Please remove or uncomment.

Reply:
Initially idea was to keep those console logs to help verify but ruins the output of tests in the terminal. Hence, removing them.